File: 35-36/go/main.go
package main

import "fmt"

func mergeSort[T comparable](array []T, comparator func(a, b T) int) []T {
	if comparator == nil {
		comparator = func(a, b T) int {
			aa := any(a).(int)
			bb := any(b).(int)
			return aa - bb
		}
	}

	var fn func(a []T) []T

	fn = func(a []T) []T {
		if len(a) <= 1 {
			return a
		}

		middle := len(a) / 2

		return merge(
			fn(a[:middle]),
			fn(a[middle:]),
			comparator,
		)
	}

	return fn(array)
}

func merge[T comparable](a, b []T, comparator func(a, b T) int) []T {
	l, r, lr, size := 0, 0, 0, len(a)+len(b)
	c := make([]T, size)

	for lr < size {
		switch {
		case l >= len(a):
			c[lr] = b[r]
			r++
		case r >= len(b):
			c[lr] = a[l]
			l++
		case comparator(a[l], b[r]) < 0:
			c[lr] = a[l]
			l++
		default:
			c[lr] = b[r]
			r++
		}
		lr++
	}

	return c
}

func main() {
	fmt.Printf("%v", merge([]int{3, 4, 5}, []int{3, 6, 7}, func(a, b int) int {
		return a - b
	}))
}

File: 35-36/py/main.py
from typing import Callable

def merge_sort[T](array: list[T], comparator: Callable[[T, T], int] | None = None) -> list[T]:
    comparator = comparator or (lambda a, b: a - b)

    def fn(array):
        if len(array) <= 1:
            return array
        
        middle = len(array) // 2

        return merge(
            fn(array[:middle]),
            fn(array[middle:]),
            comparator,
        )
     
    return fn(array)


def merge[T](a: list[T], b: list[T], comparator: Callable[[T, T], int]) -> list[T]:
    l, r, lr, size = 0, 0, 0, len(a) + len(b)
    c = [None] * size
    
    while lr < size:
        if l >= len(a):
            c[lr] = b[r]
            r += 1
        elif r >= len(b):
            c[lr] = a[l]
            l += 1
        elif comparator(a[l], b[r]) < 0:
            c[lr] = a[l]
            l += 1
        else:
            c[lr] = b[r]
            r += 1
        lr += 1
    
    return c
